In [399]:
# Import required libraries
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
%matplotlib inline

In [400]:
patients = pd.read_csv('data/insulin_trials/patients.csv')
treatments = pd.read_csv('data/insulin_trials/treatments.csv')
adverse_reactions = pd.read_csv('data/insulin_trials/adverse_reactions.csv')


In [401]:
print(patients.shape)
print(treatments.shape)
print(adverse_reactions.shape)

(503, 14)
(280, 7)
(34, 3)


In [402]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [403]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [404]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [405]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [406]:
missing_patient_details = patients[patients['address'].isnull()]

In [407]:
missing_patient_list = missing_patient_details['patient_id'].to_list()

In [408]:
missing_patient_list

[210, 220, 231, 235, 243, 250, 258, 265, 270, 279, 287, 297]

In [409]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [410]:
missing_hba1c_details = treatments[treatments['hba1c_change'].isnull()]

In [411]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


### Quality issues

['completeness', 'validity', 'accuracy', 'inconsistance']

## Patients
- zip code is in float format   [v]
- some zip codes are of four digits     [v]
- missing details for ['address, city, state, zip_code, country, contact]   [ic]
- patient with missing details ids [210, 220, 231, 235, 243, 250, 258, 265, 270, 279, 287, 297]     [ic]
- incorrect datatypes for assigned_sex and birthdate cols
- errored data for weight and heights       [a]
- inconsistant way of storing phone number in contact colm      [ic]
- duplicate and not useful data     [ic]
## Treatments
- hba1c_change col has missing values       [c]
- novodra and aurolin has 'u' attached it its numbers.      [v]
- only 280 rows instead of 350      [c]
- Name and sirname col is in lowerCase      [ic]
- incorrect data types for auralin and novadra      [a]
- hba1c col is wrongly calculated       [a]
- use of '-' instead of NaN in novodra and aurolin      [ic]


## Adverse Reaction
- No missing cols

### Tidiness

## Patients
- phone number and email are in one col         -> DONE

## Treatments
- Three variable stored in novodra and aurolin cols         -> DONE

## Adverse Reaction
- no point of using this table as the col can be integrated in 'Treatments' table           -> Merger DONE

### Completeness Issues

- missing values for contact details of 12 patients     ->NO ACTION
- only 280 rows are there out of 350                ->NO ACTION
- hba1c col has missing values      -> DONE

In [412]:
df_treatment = treatments.copy()
df_treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [413]:
df_treatment['hba1c_change_new'] = abs(df_treatment['hba1c_start']- df_treatment['hba1c_end'])

In [414]:
df_treatment.head()


given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  hba1c_change_new  
0           NaN              0.43  
1          0.97              0.47  
2           NaN              0.43  
3          0.35              0.35  
4          0.32              0.32

In [415]:
df_treatment['hba1c_change'] = df_treatment['hba1c_change_new']
df_treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  hba1c_change_new  
0          0.43              0.43  
1          0.47              0.47  
2          0.43              0.43  
3          0.35              0.35  
4          0.32              0.32

In [416]:
df_treatment.drop('hba1c_change_new', axis=1, inplace=True)
df_treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [417]:
print(df_treatment['hba1c_change'].isnull().sum())
df_treatment.describe()

0


hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    280.000000
mean      7.985929    7.589286      0.396643
std       0.568638    0.569672      0.060168
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.350000
50%       7.800000    7.420000      0.390000
75%       7.970000    7.570000      0.440000
max       9.950000    9.580000      0.540000

### Tidiness

- phone+email in single col in patient df                   -> DONE
- novodra and aurolin col treatment                   -> DONE
- merge the adverse table to treatment table        -> DONE

In [418]:
patient_df = patients.copy()
patient_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [419]:
patient_df['contact']

patient_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [420]:
phone_pattern = r'(\+?\d[\d\s\-\(\)]{6,}\d)'   # matches phone numbers with +, digits, spaces, dashes, parentheses
email_pattern = r'([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})'  # matches emails

patient_df['phone_number'] = patient_df['contact'].apply(lambda x: re.search(phone_pattern,str(x)))
patient_df['phone_number'] = patient_df['phone_number'].apply(lambda x: x.group(1) if x else '')

patient_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number  
0  19.6       951-719-9170  
1  19.2  +1 (217) 569-3204  
2  24.8       402-363-6804  
3  31.7  +1 (732) 636-8246  
4  26.1       334-515-7487

In [421]:
email_pattern = r'([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})'

# Extract email
patient_df['email_id'] = patient_df['contact'].str.extract(email_pattern)

# Check
patient_df[['contact', 'phone_number', 'email_id']].head()

contact       phone_number  \
0        951-719-9170ZoeWellish@superrito.com       951-719-9170   
1       PamelaSHill@cuvox.de+1 (217) 569-3204  +1 (217) 569-3204   
2            402-363-6804JaeMDebord@gustr.com       402-363-6804   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  +1 (732) 636-8246   
4             334-515-7487TimNeudorf@cuvox.de       334-515-7487   

                               email_id  
0  951-719-9170ZoeWellish@superrito.com  
1                  PamelaSHill@cuvox.de  
2      402-363-6804JaeMDebord@gustr.com  
3             PhanBaLiem@jourrapide.com  
4       334-515-7487TimNeudorf@cuvox.de

In [422]:
patient_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number                              email_id  
0  19.6       951-719-9170  951-719-9170ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204                  PamelaSHill@cuvox.de  
2  24.8       402-363-6804      402-363-6804JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246             PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487       334-515-7487TimNeudorf@cuvox.de

In [423]:
def clean_email(row):
    phone = str(row['phone_number'])
    email = row['email_id']

    if pd.notnull(phone) and pd.notnull(email):
        return email.replace(phone, '').strip()
    else:
        return email
    
patient_df['email_id'] = patient_df.apply(clean_email, axis=1)

In [424]:
patient_df[['contact', 'phone_number', 'email_id']].head()

contact       phone_number  \
0        951-719-9170ZoeWellish@superrito.com       951-719-9170   
1       PamelaSHill@cuvox.de+1 (217) 569-3204  +1 (217) 569-3204   
2            402-363-6804JaeMDebord@gustr.com       402-363-6804   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  +1 (732) 636-8246   
4             334-515-7487TimNeudorf@cuvox.de       334-515-7487   

                    email_id  
0   ZoeWellish@superrito.com  
1       PamelaSHill@cuvox.de  
2       JaeMDebord@gustr.com  
3  PhanBaLiem@jourrapide.com  
4        TimNeudorf@cuvox.de

In [425]:
patient_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number                   email_id  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

In [426]:
patient_df.drop('contact', axis=1, inplace=True)
patient_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi       phone_number                   email_id  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

In [427]:
df_treatment.columns

Index(['given_name', 'surname', 'auralin', 'novodra', 'hba1c_start',
       'hba1c_end', 'hba1c_change'],
      dtype='object')

In [428]:
df_treatment = pd.melt(df_treatment, id_vars=['given_name', 'surname', 'hba1c_start','hba1c_end', 'hba1c_change'],
                       var_name='treatment', value_name='dose')

In [429]:
df_treatment =df_treatment[df_treatment['dose']!= '-']
df_treatment

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika     jindrová         7.63       7.20          0.43   auralin   
3         skye   gormanston         7.97       7.62          0.35   auralin   
6       sophia       haugen         7.65       7.27          0.38   auralin   
7        eddie       archer         7.89       7.55          0.34   auralin   
9         asia      woźniak         7.76       7.37          0.39   auralin   
..         ...          ...          ...        ...           ...       ...   
548     yasmin       araujo         7.82       7.36          0.46   novodra   
549     hiromu     horikawa         7.77       7.28          0.49   novodra   
551        leo       vieira         7.74       7.36          0.38   novodra   
552     steven          roy         7.87       7.43          0.44   novodra   
556       john  teichelmann         7.90       7.58          0.32   novodra   

          dose  
0    41u - 48u  
3    33u - 36u  
6    37u - 42u  
7    31u - 38u  
9    30u - 36u  
..         ...  
548  51u - 54u  
549  47u - 46u  
551  30u - 33u  
552  41u - 43u  
556  49u - 49u  

[280 rows x 7 columns]

In [430]:
df_treatment[['start', 'end']] = df_treatment['dose'].str.split('-', expand=True)

In [431]:
df_treatment

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika     jindrová         7.63       7.20          0.43   auralin   
3         skye   gormanston         7.97       7.62          0.35   auralin   
6       sophia       haugen         7.65       7.27          0.38   auralin   
7        eddie       archer         7.89       7.55          0.34   auralin   
9         asia      woźniak         7.76       7.37          0.39   auralin   
..         ...          ...          ...        ...           ...       ...   
548     yasmin       araujo         7.82       7.36          0.46   novodra   
549     hiromu     horikawa         7.77       7.28          0.49   novodra   
551        leo       vieira         7.74       7.36          0.38   novodra   
552     steven          roy         7.87       7.43          0.44   novodra   
556       john  teichelmann         7.90       7.58          0.32   novodra   

          dose start   end  
0    41u - 48u  41u    48u  
3    33u - 36u  33u    36u  
6    37u - 42u  37u    42u  
7    31u - 38u  31u    38u  
9    30u - 36u  30u    36u  
..         ...   ...   ...  
548  51u - 54u  51u    54u  
549  47u - 46u  47u    46u  
551  30u - 33u  30u    33u  
552  41u - 43u  41u    43u  
556  49u - 49u  49u    49u  

[280 rows x 9 columns]

In [432]:
df_treatment.drop('dose', axis=1, inplace=True)
df_treatment

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika     jindrová         7.63       7.20          0.43   auralin   
3         skye   gormanston         7.97       7.62          0.35   auralin   
6       sophia       haugen         7.65       7.27          0.38   auralin   
7        eddie       archer         7.89       7.55          0.34   auralin   
9         asia      woźniak         7.76       7.37          0.39   auralin   
..         ...          ...          ...        ...           ...       ...   
548     yasmin       araujo         7.82       7.36          0.46   novodra   
549     hiromu     horikawa         7.77       7.28          0.49   novodra   
551        leo       vieira         7.74       7.36          0.38   novodra   
552     steven          roy         7.87       7.43          0.44   novodra   
556       john  teichelmann         7.90       7.58          0.32   novodra   

    start   end  
0    41u    48u  
3    33u    36u  
6    37u    42u  
7    31u    38u  
9    30u    36u  
..    ...   ...  
548  51u    54u  
549  47u    46u  
551  30u    33u  
552  41u    43u  
556  49u    49u  

[280 rows x 8 columns]

In [433]:
df_treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

In [434]:
df_adverse_reactions = adverse_reactions.copy()

In [435]:
df_treatment = df_treatment.merge(
    df_adverse_reactions,
    on=['given_name', 'surname'],
    how='left'
)


In [436]:
df_treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end adverse_reaction  
0  41u    48u              NaN  
1  33u    36u              NaN  
2  37u    42u              NaN  
3  31u    38u              NaN  
4  30u    36u              NaN

In [437]:
df_treatment['adverse_reaction'] = df_treatment['adverse_reaction'].fillna('No reaction')

df_treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end adverse_reaction  
0  41u    48u      No reaction  
1  33u    36u      No reaction  
2  37u    42u      No reaction  
3  31u    38u      No reaction  
4  30u    36u      No reaction

In [438]:
df_treatment['adverse_reaction'].value_counts()

adverse_reaction
No reaction                  253
hypoglycemia                  18
headache                       3
throat irritation              2
injection site discomfort      2
nausea                         1
cough                          1
Name: count, dtype: int64

### Quality issues

['completeness', 'validity', 'accuracy', 'inconsistance']

## Patients
- zip code is in float format   [v]         -> DONE
- some zip codes are of four digits     [v]         -> DONE
-
-
- incorrect datatypes for assigned_sex and birthdate cols  -> DONE
- errored data for weight and heights       [a]
-
- duplicate and not useful data  (JOHN DOE)   [ic]      -> DONE

## Treatments
- 
- novodra and aurolin has 'u' attached it its numbers.      [v]
- 
- Name and surname col is in lowerCase      [ic]
- incorrect data types for auralin and novadra      [a]
- 
- 

In [439]:
patient_df['zip_code'].fillna(0.0, inplace=True)

/var/folders/sv/g8vyvw8s2136bxp6jkg5vx3w0000gn/T/ipykernel_73098/2345606824.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  patient_df['zip_code'].fillna(0.0, inplace=True)


In [440]:

patient_df['zip_code'] = patient_df['zip_code'].astype(str).str.replace(r'\.0$', '', regex=True).str.zfill(5)


In [441]:
patient_df

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state zip_code  \
0         576 Brown Bear Drive  Rancho California  California    92390   
1    2370 University Hill Road          Armstrong    Illinois    61812   
2        1493 Poling Farm Road               York    Nebraska    68467   
3          2335 Webster Street         Woodbridge          NJ    07095   
4         1428 Turkey Pen Lane             Dothan          AL    36303   
..                         ...                ...         ...      ...   
498        2530 Victoria Court       Milton Mills          ME    03852   
499   494 Clarksburg Park Road             Sedona          AZ    86341   
500          649 Nutter Street      Overland Park          MO    64110   
501   3652 Boone Crockett Lane            Seattle          WA    98109   
502          2778 North Avenue               Burr    Nebraska    68324   

           country  birthdate  weight  height   bmi       phone_number  \
0    United States  7/10/1976   121.7      66  19.6       951-719-9170   
1    United States   4/3/1967   118.8      66  19.2  +1 (217) 569-3204   
2    United States  2/19/1980   177.8      71  24.8       402-363-6804   
3    United States  7/26/1951   220.9      70  31.7  +1 (732) 636-8246   
4    United States  2/18/1928   192.3      27  26.1       334-515-7487   
..             ...        ...     ...     ...   ...                ...   
498  United States  4/10/1959   181.1      72  24.6       207-477-0579   
499  United States  3/26/1948   239.6      70  34.4       928-284-4492   
500  United States  1/13/1971   171.2      67  26.8       816-223-6007   
501  United States  2/13/1952   176.9      67  27.7     1 360 443 2060   
502  United States   5/3/1954   138.2      71  19.3       402-848-4923   

                              email_id  
0             ZoeWellish@superrito.com  
1                 PamelaSHill@cuvox.de  
2                 JaeMDebord@gustr.com  
3            PhanBaLiem@jourrapide.com  
4                  TimNeudorf@cuvox.de  
..                                 ...  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com  

[503 rows x 15 columns]

In [442]:
patient_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      503 non-null    object 
 8   country       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
 13  phone_number  503 non-null    object 
 14  email_id      491 non-null    object 
dtypes: float64(2), int64(2), object(11)
memory usage: 59.1+ KB


In [443]:
patient_df['assigned_sex'] = patient_df['assigned_sex'].astype('category')
patient_df.describe()

patient_id      weight      height         bmi
count  503.000000  503.000000  503.000000  503.000000
mean   252.000000  173.434990   66.634195   27.483897
std    145.347859   33.916741    4.411297    5.276438
min      1.000000   48.800000   27.000000   17.100000
25%    126.500000  149.300000   63.000000   23.300000
50%    252.000000  175.300000   67.000000   27.200000
75%    377.500000  199.500000   70.000000   31.750000
max    503.000000  255.900000   79.000000   37.700000

In [444]:
patient_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi       phone_number                   email_id  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

In [445]:
john_doe_index = patient_df[(patient_df['given_name'] == 'John') & (patient_df['surname']=='Doe')].index


In [446]:
len(john_doe_index)

6

In [447]:
patient_df.drop(index=john_doe_index, inplace=True)


In [448]:
patient_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 497 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   patient_id    497 non-null    int64   
 1   assigned_sex  497 non-null    category
 2   given_name    497 non-null    object  
 3   surname       497 non-null    object  
 4   address       485 non-null    object  
 5   city          485 non-null    object  
 6   state         485 non-null    object  
 7   zip_code      497 non-null    object  
 8   country       485 non-null    object  
 9   birthdate     497 non-null    object  
 10  weight        497 non-null    float64 
 11  height        497 non-null    int64   
 12  bmi           497 non-null    float64 
 13  phone_number  497 non-null    object  
 14  email_id      485 non-null    object  
dtypes: category(1), float64(2), int64(2), object(10)
memory usage: 58.8+ KB


In [449]:
df_missing_details =  patient_df[patient_df['address'].isnull()][['given_name', 'surname']]

In [450]:
df_missing_details

given_name      surname
209     Lalita  Eldarkhanov
219         Mỹ        Quynh
230  Elisabeth      Knudsen
234    Martina    Tománková
242       John      O'Brian
249   Benjamin       Mehler
257        Jin         Kung
264   Wafiyyah       Asfour
269     Flavia   Fiorentino
278   Generosa        Cabán
286      Lewis         Webb
296        Chỉ          Lâm

In [451]:
df_common = df_missing_details.merge(
    df_treatment,
    on=['given_name', 'surname'],
    how='inner'
)



In [452]:
df_common

# THE names without details in patient table are not present in treatments table 
# Therefore, its ok to remove them from our patients table

Empty DataFrame
Columns: [given_name, surname, hba1c_start, hba1c_end, hba1c_change, treatment, start, end, adverse_reaction]
Index: []

In [453]:
patient_df.dropna(subset=['address'], inplace=True)


In [454]:
patient_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 485 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   patient_id    485 non-null    int64   
 1   assigned_sex  485 non-null    category
 2   given_name    485 non-null    object  
 3   surname       485 non-null    object  
 4   address       485 non-null    object  
 5   city          485 non-null    object  
 6   state         485 non-null    object  
 7   zip_code      485 non-null    object  
 8   country       485 non-null    object  
 9   birthdate     485 non-null    object  
 10  weight        485 non-null    float64 
 11  height        485 non-null    int64   
 12  bmi           485 non-null    float64 
 13  phone_number  485 non-null    object  
 14  email_id      485 non-null    object  
dtypes: category(1), float64(2), int64(2), object(10)
memory usage: 57.4+ KB


## Treatments
- 
- novodra and aurolin has 'u' attached it its numbers.      [v]         -> DONE
- 
- Name and surname col is in lowerCase      [ic]            -> DONE
- incorrect data types for auralin and novadra      [a]     -> DONE
- 

In [455]:
df_treatment['given_name'] = df_treatment['given_name'].str.capitalize()
df_treatment['surname'] = df_treatment['surname'].str.capitalize()

df_treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   Veronika    Jindrová         7.63       7.20          0.43   auralin   
1       Skye  Gormanston         7.97       7.62          0.35   auralin   
2     Sophia      Haugen         7.65       7.27          0.38   auralin   
3      Eddie      Archer         7.89       7.55          0.34   auralin   
4       Asia     Woźniak         7.76       7.37          0.39   auralin   

  start   end adverse_reaction  
0  41u    48u      No reaction  
1  33u    36u      No reaction  
2  37u    42u      No reaction  
3  31u    38u      No reaction  
4  30u    36u      No reaction

In [456]:
df_treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   given_name        280 non-null    object 
 1   surname           280 non-null    object 
 2   hba1c_start       280 non-null    float64
 3   hba1c_end         280 non-null    float64
 4   hba1c_change      280 non-null    float64
 5   treatment         280 non-null    object 
 6   start             280 non-null    object 
 7   end               280 non-null    object 
 8   adverse_reaction  280 non-null    object 
dtypes: float64(3), object(6)
memory usage: 19.8+ KB


In [457]:
df_treatment['start'] = df_treatment['start'].str.replace('u','', regex=False).astype(int)
df_treatment['end'] = df_treatment['end'].str.replace('u','', regex=False).astype(int)
df_treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   Veronika    Jindrová         7.63       7.20          0.43   auralin   
1       Skye  Gormanston         7.97       7.62          0.35   auralin   
2     Sophia      Haugen         7.65       7.27          0.38   auralin   
3      Eddie      Archer         7.89       7.55          0.34   auralin   
4       Asia     Woźniak         7.76       7.37          0.39   auralin   

   start  end adverse_reaction  
0     41   48      No reaction  
1     33   36      No reaction  
2     37   42      No reaction  
3     31   38      No reaction  
4     30   36      No reaction

In [459]:
df_treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   given_name        280 non-null    object 
 1   surname           280 non-null    object 
 2   hba1c_start       280 non-null    float64
 3   hba1c_end         280 non-null    float64
 4   hba1c_change      280 non-null    float64
 5   treatment         280 non-null    object 
 6   start             280 non-null    int64  
 7   end               280 non-null    int64  
 8   adverse_reaction  280 non-null    object 
dtypes: float64(3), int64(2), object(4)
memory usage: 19.8+ KB


In [460]:
patient_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 485 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   patient_id    485 non-null    int64   
 1   assigned_sex  485 non-null    category
 2   given_name    485 non-null    object  
 3   surname       485 non-null    object  
 4   address       485 non-null    object  
 5   city          485 non-null    object  
 6   state         485 non-null    object  
 7   zip_code      485 non-null    object  
 8   country       485 non-null    object  
 9   birthdate     485 non-null    object  
 10  weight        485 non-null    float64 
 11  height        485 non-null    int64   
 12  bmi           485 non-null    float64 
 13  phone_number  485 non-null    object  
 14  email_id      485 non-null    object  
dtypes: category(1), float64(2), int64(2), object(10)
memory usage: 57.4+ KB


# OUR BOTH THE TABLE'S DATA IS CLEANED

In [461]:
patient_df.to_csv('data/insulin_trials/cleaned_patients_data.csv')

In [462]:
df_treatment.to_csv('data/insulin_trials/cleaned_treatments_data.csv')